src/redux/books/booksSlice.js

```diff
@@ -1,40 +1,103 @@
+// ASYNC ACTIONS
+const fetchBooks = createAsyncThunk('DisplayBooks', async () => {
+  const response = await axios.get(API_URL);
+  return response.data;
+});
+
+const addBook = createAsyncThunk('AddBook', async (book) => {
+  const response = await axios.post(API_URL, book);
+  return response.data === 'Created' ? book : null;
+});
+
+const removeBook = createAsyncThunk('RemoveBook', async (ITEM_ID) => {
+  const response = await axios.delete(`${API_URL}/${ITEM_ID}`);
+  return response.data === 'The book is removed!' ? ITEM_ID : null;
+});
+
+// INITIAL STATE
+const initialState = {
+  books: [],
+  error: '',
+  status: 'idle',
+};
+
+// SLICE
+export const booksSlice = createSlice({
   name: 'books',
-  initialState: initialBooksState,
-  reducers: {
-    addBook: (state, action) => {
-      state.push(action.payload);
-    },
-    removeBook: (state, action) => (
-      state.filter((book) => book.item_id !== action.payload)
-    ),
+  initialState,
+  reducers: {},
+  extraReducers: (builder) => {
+    builder
+      .addCase(fetchBooks.pending, (state) => {
+        state.status = 'loading';
+      })
+      .addCase(fetchBooks.fulfilled, (state, action) => {
+        state.status = 'fetched';
+        if (action.payload !== '') {
+          const books = [];
+          const keys = Object.keys(action.payload);
+          keys.forEach((Id) => {
+            books.push({ item_id: Id, ...action.payload[Id][0] });
+          });
+          state.books = books;
+          if (state.books.length === 0) {
+            state.error = 'No books found';
+          }
+        } else {
+          state.error = 'No books found';
+        }
+      })
+      .addCase(fetchBooks.rejected, (state, action) => {
+        state.status = 'failed';
+        state.error = action.error.message;
+      });
+
+    builder
+      .addCase(addBook.pending, (state) => {
+        state.status = 'pending';
+      })
+      .addCase(addBook.fulfilled, (state, action) => {
+        if (action.payload !== null) {
+          state.status = 'succeeded';
+          state.error = '';
+          state.books.push(action.payload);
+        } else {
+          state.status = 'failed';
+          state.error = 'Unable to add book';
+        }
+      })
+      .addCase(addBook.rejected, (state, action) => {
+        state.status = 'failed';
+        state.error = action.error.message;
+      });
+
+    builder
+      .addCase(removeBook.pending, (state) => {
+        state.status = 'loading';
+      })
+      .addCase(removeBook.fulfilled, (state, action) => {
+        if (action.payload !== null) {
+          state.status = 'succeeded';
+          state.error = '';
+          state.books = state.books.filter((bookId) => bookId.item_id !== action.payload);
+          if (state.books.length === 0) state.error = 'No book was found!';
+        } else {
+          state.status = 'failed';
+          state.error = 'Unable to remove book';
+        }
+      })
```

Comment:
- [ ] Great job on the refactor of your remove book functionality, and kudos for successfully removing the book from the API! 📚👏 However, during my testing, I noticed an issue that I'd like to bring to your attention. 

  - 🚀 While you've done an impressive job, there's a small hiccup: When I click on the `remove` button, the book gets removed from the `API`, but the `UI` doesn't reflect the current state of the books in your `API`. This kind of discrepancy often occurs when the UI doesn't update in response to actions.

  - To ensure a seamless user experience, please consider updating the UI's state of books whenever the remove book button is triggered. This will help maintain consistency between your application's front end and back end. 🔄💼

Happy coding! 😊🚀🍀

Reply:
Thanks